{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e76652ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install pandas numpy scikit-learn xgboost joblib biopython optuna shap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c0f4c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cell 1\n",
    "import os\n",
    "import math\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from collections import Counter, defaultdict\n",
    "from Bio.Seq import Seq\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import accuracy_score, classification_report, confusion_matrix\n",
    "from xgboost import XGBClassifier\n",
    "import joblib\n",
    "import optuna\n",
    "import shap\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7177fe8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cell 2\n",
    "CSV_PATH = \"codon_usage.csv\"   \n",
    "assert os.path.exists(CSV_PATH), f\"Put your Kaggle CSV at: {CSV_PATH}\"\n",
    "\n",
    "df_raw = pd.read_csv(CSV_PATH)\n",
    "print(\"Loaded shape:\", df_raw.shape)\n",
    "display(df_raw.head())\n",
    "print(\"\\nColumns:\", list(df_raw.columns))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "35fcb4d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded: codon_usage.csv\n",
      "Shape: (13028, 69)\n",
      "Columns:\n",
      "  1. 'Kingdom'\n",
      "  2. 'DNAtype'\n",
      "  3. 'SpeciesID'\n",
      "  4. 'Ncodons'\n",
      "  5. 'SpeciesName'\n",
      "  6. 'UUU'\n",
      "  7. 'UUC'\n",
      "  8. 'UUA'\n",
      "  9. 'UUG'\n",
      "  10. 'CUU'\n",
      "  11. 'CUC'\n",
      "  12. 'CUA'\n",
      "  13. 'CUG'\n",
      "  14. 'AUU'\n",
      "  15. 'AUC'\n",
      "  16. 'AUA'\n",
      "  17. 'AUG'\n",
      "  18. 'GUU'\n",
      "  19. 'GUC'\n",
      "  20. 'GUA'\n",
      "  21. 'GUG'\n",
      "  22. 'GCU'\n",
      "  23. 'GCC'\n",
      "  24. 'GCA'\n",
      "  25. 'GCG'\n",
      "  26. 'CCU'\n",
      "  27. 'CCC'\n",
      "  28. 'CCA'\n",
      "  29. 'CCG'\n",
      "  30. 'UGG'\n",
      "  31. 'GGU'\n",
      "  32. 'GGC'\n",
      "  33. 'GGA'\n",
      "  34. 'GGG'\n",
      "  35. 'UCU'\n",
      "  36. 'UCC'\n",
      "  37. 'UCA'\n",
      "  38. 'UCG'\n",
      "  39. 'AGU'\n",
      "  40. 'AGC'\n",
      "  41. 'ACU'\n",
      "  42. 'ACC'\n",
      "  43. 'ACA'\n",
      "  44. 'ACG'\n",
      "  45. 'UAU'\n",
      "  46. 'UAC'\n",
      "  47. 'CAA'\n",
      "  48. 'CAG'\n",
      "  49. 'AAU'\n",
      "  50. 'AAC'\n",
      "  51. 'UGU'\n",
      "  52. 'UGC'\n",
      "  53. 'CAU'\n",
      "  54. 'CAC'\n",
      "  55. 'AAA'\n",
      "  56. 'AAG'\n",
      "  57. 'CGU'\n",
      "  58. 'CGC'\n",
      "  59. 'CGA'\n",
      "  60. 'CGG'\n",
      "  61. 'AGA'\n",
      "  62. 'AGG'\n",
      "  63. 'GAU'\n",
      "  64. 'GAC'\n",
      "  65. 'GAA'\n",
      "  66. 'GAG'\n",
      "  67. 'UAA'\n",
      "  68. 'UAG'\n",
      "  69. 'UGA'\n",
      "\n",
      "First 8 rows preview:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>DNAtype</th>\n",
       "      <th>SpeciesID</th>\n",
       "      <th>Ncodons</th>\n",
       "      <th>SpeciesName</th>\n",
       "      <th>UUU</th>\n",
       "      <th>UUC</th>\n",
       "      <th>UUA</th>\n",
       "      <th>UUG</th>\n",
       "      <th>CUU</th>\n",
       "      <th>CUC</th>\n",
       "      <th>CUA</th>\n",
       "      <th>CUG</th>\n",
       "      <th>AUU</th>\n",
       "      <th>AUC</th>\n",
       "      <th>AUA</th>\n",
       "      <th>AUG</th>\n",
       "      <th>GUU</th>\n",
       "      <th>GUC</th>\n",
       "      <th>GUA</th>\n",
       "      <th>GUG</th>\n",
       "      <th>GCU</th>\n",
       "      <th>GCC</th>\n",
       "      <th>GCA</th>\n",
       "      <th>GCG</th>\n",
       "      <th>CCU</th>\n",
       "      <th>CCC</th>\n",
       "      <th>CCA</th>\n",
       "      <th>CCG</th>\n",
       "      <th>UGG</th>\n",
       "      <th>GGU</th>\n",
       "      <th>GGC</th>\n",
       "      <th>GGA</th>\n",
       "      <th>GGG</th>\n",
       "      <th>UCU</th>\n",
       "      <th>UCC</th>\n",
       "      <th>UCA</th>\n",
       "      <th>UCG</th>\n",
       "      <th>AGU</th>\n",
       "      <th>AGC</th>\n",
       "      <th>ACU</th>\n",
       "      <th>ACC</th>\n",
       "      <th>ACA</th>\n",
       "      <th>ACG</th>\n",
       "      <th>UAU</th>\n",
       "      <th>UAC</th>\n",
       "      <th>CAA</th>\n",
       "      <th>CAG</th>\n",
       "      <th>AAU</th>\n",
       "      <th>AAC</th>\n",
       "      <th>UGU</th>\n",
       "      <th>UGC</th>\n",
       "      <th>CAU</th>\n",
       "      <th>CAC</th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAG</th>\n",
       "      <th>CGU</th>\n",
       "      <th>CGC</th>\n",
       "      <th>CGA</th>\n",
       "      <th>CGG</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGG</th>\n",
       "      <th>GAU</th>\n",
       "      <th>GAC</th>\n",
       "      <th>GAA</th>\n",
       "      <th>GAG</th>\n",
       "      <th>UAA</th>\n",
       "      <th>UAG</th>\n",
       "      <th>UGA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>vrl</td>\n",
       "      <td>0</td>\n",
       "      <td>100217</td>\n",
       "      <td>1995</td>\n",
       "      <td>Epizootic haematopoietic necrosis virus</td>\n",
       "      <td>0.01654</td>\n",
       "      <td>0.01203</td>\n",
       "      <td>5e-04</td>\n",
       "      <td>0.00351</td>\n",
       "      <td>0.01203</td>\n",
       "      <td>0.03208</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.0401</td>\n",
       "      <td>0.00551</td>\n",
       "      <td>0.02005</td>\n",
       "      <td>0.00752</td>\n",
       "      <td>0.02506</td>\n",
       "      <td>0.01103</td>\n",
       "      <td>0.0411</td>\n",
       "      <td>0.00902</td>\n",
       "      <td>0.03308</td>\n",
       "      <td>0.01003</td>\n",
       "      <td>0.05013</td>\n",
       "      <td>0.01554</td>\n",
       "      <td>0.01103</td>\n",
       "      <td>0.02356</td>\n",
       "      <td>0.03208</td>\n",
       "      <td>0.01203</td>\n",
       "      <td>0.00501</td>\n",
       "      <td>0.01003</td>\n",
       "      <td>0.01203</td>\n",
       "      <td>0.03158</td>\n",
       "      <td>0.01905</td>\n",
       "      <td>0.02456</td>\n",
       "      <td>0.01353</td>\n",
       "      <td>0.02155</td>\n",
       "      <td>0.00251</td>\n",
       "      <td>0.00652</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0.01554</td>\n",
       "      <td>0.00501</td>\n",
       "      <td>0.02105</td>\n",
       "      <td>0.00902</td>\n",
       "      <td>0.01053</td>\n",
       "      <td>0.00501</td>\n",
       "      <td>0.02256</td>\n",
       "      <td>0.00301</td>\n",
       "      <td>0.03108</td>\n",
       "      <td>0.00401</td>\n",
       "      <td>0.02607</td>\n",
       "      <td>0.00251</td>\n",
       "      <td>0.01153</td>\n",
       "      <td>0.00501</td>\n",
       "      <td>0.02356</td>\n",
       "      <td>0.01053</td>\n",
       "      <td>0.0386</td>\n",
       "      <td>0.00401</td>\n",
       "      <td>0.00702</td>\n",
       "      <td>0.00401</td>\n",
       "      <td>0.00451</td>\n",
       "      <td>0.01303</td>\n",
       "      <td>0.03559</td>\n",
       "      <td>0.01003</td>\n",
       "      <td>0.04612</td>\n",
       "      <td>0.01203</td>\n",
       "      <td>0.04361</td>\n",
       "      <td>0.00251</td>\n",
       "      <td>5e-04</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>vrl</td>\n",
       "      <td>0</td>\n",
       "      <td>100220</td>\n",
       "      <td>1474</td>\n",
       "      <td>Bohle iridovirus</td>\n",
       "      <td>0.02714</td>\n",
       "      <td>0.01357</td>\n",
       "      <td>0.00068</td>\n",
       "      <td>0.00678</td>\n",
       "      <td>0.00407</td>\n",
       "      <td>0.02849</td>\n",
       "      <td>0.00204</td>\n",
       "      <td>0.0441</td>\n",
       "      <td>0.01153</td>\n",
       "      <td>0.0251</td>\n",
       "      <td>0.00882</td>\n",
       "      <td>0.03324</td>\n",
       "      <td>0.00814</td>\n",
       "      <td>0.04071</td>\n",
       "      <td>0.00814</td>\n",
       "      <td>0.03256</td>\n",
       "      <td>0.01085</td>\n",
       "      <td>0.04885</td>\n",
       "      <td>0.01221</td>\n",
       "      <td>0.01357</td>\n",
       "      <td>0.00678</td>\n",
       "      <td>0.02714</td>\n",
       "      <td>0.01221</td>\n",
       "      <td>0.00407</td>\n",
       "      <td>0.01425</td>\n",
       "      <td>0.01221</td>\n",
       "      <td>0.01967</td>\n",
       "      <td>0.02239</td>\n",
       "      <td>0.01289</td>\n",
       "      <td>0.02103</td>\n",
       "      <td>0.01493</td>\n",
       "      <td>0.00407</td>\n",
       "      <td>0.00475</td>\n",
       "      <td>0.00068</td>\n",
       "      <td>0.02035</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>0.02782</td>\n",
       "      <td>0.01425</td>\n",
       "      <td>0.00611</td>\n",
       "      <td>0.00475</td>\n",
       "      <td>0.02917</td>\n",
       "      <td>0.00407</td>\n",
       "      <td>0.02374</td>\n",
       "      <td>0.00882</td>\n",
       "      <td>0.02917</td>\n",
       "      <td>0.00271</td>\n",
       "      <td>0.01628</td>\n",
       "      <td>0.00204</td>\n",
       "      <td>0.01967</td>\n",
       "      <td>0.00543</td>\n",
       "      <td>0.03392</td>\n",
       "      <td>0.00136</td>\n",
       "      <td>0.00678</td>\n",
       "      <td>0.00136</td>\n",
       "      <td>0.00136</td>\n",
       "      <td>0.01696</td>\n",
       "      <td>0.03596</td>\n",
       "      <td>0.01221</td>\n",
       "      <td>0.04545</td>\n",
       "      <td>0.0156</td>\n",
       "      <td>0.0441</td>\n",
       "      <td>0.00271</td>\n",
       "      <td>0.00068</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>vrl</td>\n",
       "      <td>0</td>\n",
       "      <td>100755</td>\n",
       "      <td>4862</td>\n",
       "      <td>Sweet potato leaf curl virus</td>\n",
       "      <td>0.01974</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.01357</td>\n",
       "      <td>0.01543</td>\n",
       "      <td>0.00782</td>\n",
       "      <td>0.01111</td>\n",
       "      <td>0.01028</td>\n",
       "      <td>0.01193</td>\n",
       "      <td>0.02283</td>\n",
       "      <td>0.01604</td>\n",
       "      <td>0.01316</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.01625</td>\n",
       "      <td>0.01872</td>\n",
       "      <td>0.01213</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0.02406</td>\n",
       "      <td>0.01234</td>\n",
       "      <td>0.0144</td>\n",
       "      <td>0.00514</td>\n",
       "      <td>0.01604</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.02098</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0.01728</td>\n",
       "      <td>0.01851</td>\n",
       "      <td>0.00864</td>\n",
       "      <td>0.01172</td>\n",
       "      <td>0.01892</td>\n",
       "      <td>0.01933</td>\n",
       "      <td>0.01419</td>\n",
       "      <td>0.01296</td>\n",
       "      <td>0.00967</td>\n",
       "      <td>0.01337</td>\n",
       "      <td>0.01337</td>\n",
       "      <td>0.01851</td>\n",
       "      <td>0.01131</td>\n",
       "      <td>0.01419</td>\n",
       "      <td>0.0109</td>\n",
       "      <td>0.02612</td>\n",
       "      <td>0.01275</td>\n",
       "      <td>0.01522</td>\n",
       "      <td>0.02365</td>\n",
       "      <td>0.02962</td>\n",
       "      <td>0.01789</td>\n",
       "      <td>0.01625</td>\n",
       "      <td>0.01234</td>\n",
       "      <td>0.01604</td>\n",
       "      <td>0.01687</td>\n",
       "      <td>0.02077</td>\n",
       "      <td>0.03949</td>\n",
       "      <td>0.00864</td>\n",
       "      <td>0.00596</td>\n",
       "      <td>0.00926</td>\n",
       "      <td>0.00596</td>\n",
       "      <td>0.01974</td>\n",
       "      <td>0.02489</td>\n",
       "      <td>0.03126</td>\n",
       "      <td>0.02036</td>\n",
       "      <td>0.02242</td>\n",
       "      <td>0.02468</td>\n",
       "      <td>0.00391</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>vrl</td>\n",
       "      <td>0</td>\n",
       "      <td>100880</td>\n",
       "      <td>1915</td>\n",
       "      <td>Northern cereal mosaic virus</td>\n",
       "      <td>0.01775</td>\n",
       "      <td>0.02245</td>\n",
       "      <td>0.01619</td>\n",
       "      <td>0.00992</td>\n",
       "      <td>0.01567</td>\n",
       "      <td>0.01358</td>\n",
       "      <td>0.0094</td>\n",
       "      <td>0.01723</td>\n",
       "      <td>0.02402</td>\n",
       "      <td>0.02245</td>\n",
       "      <td>0.02507</td>\n",
       "      <td>0.02924</td>\n",
       "      <td>0.02089</td>\n",
       "      <td>0.02141</td>\n",
       "      <td>0.01723</td>\n",
       "      <td>0.01932</td>\n",
       "      <td>0.02141</td>\n",
       "      <td>0.00679</td>\n",
       "      <td>0.02245</td>\n",
       "      <td>0.00522</td>\n",
       "      <td>0.01358</td>\n",
       "      <td>0.00418</td>\n",
       "      <td>0.0141</td>\n",
       "      <td>0.00574</td>\n",
       "      <td>0.01201</td>\n",
       "      <td>0.00992</td>\n",
       "      <td>0.00366</td>\n",
       "      <td>0.02402</td>\n",
       "      <td>0.02663</td>\n",
       "      <td>0.02872</td>\n",
       "      <td>0.00992</td>\n",
       "      <td>0.0235</td>\n",
       "      <td>0.00522</td>\n",
       "      <td>0.01619</td>\n",
       "      <td>0.00836</td>\n",
       "      <td>0.02037</td>\n",
       "      <td>0.01358</td>\n",
       "      <td>0.02089</td>\n",
       "      <td>0.00731</td>\n",
       "      <td>0.02141</td>\n",
       "      <td>0.00888</td>\n",
       "      <td>0.01567</td>\n",
       "      <td>0.01253</td>\n",
       "      <td>0.02298</td>\n",
       "      <td>0.01358</td>\n",
       "      <td>0.00992</td>\n",
       "      <td>0.00888</td>\n",
       "      <td>0.00783</td>\n",
       "      <td>0.00679</td>\n",
       "      <td>0.03133</td>\n",
       "      <td>0.04282</td>\n",
       "      <td>0.00627</td>\n",
       "      <td>0.00261</td>\n",
       "      <td>0.00261</td>\n",
       "      <td>0.00366</td>\n",
       "      <td>0.0141</td>\n",
       "      <td>0.01671</td>\n",
       "      <td>0.0376</td>\n",
       "      <td>0.01932</td>\n",
       "      <td>0.03029</td>\n",
       "      <td>0.03446</td>\n",
       "      <td>0.00261</td>\n",
       "      <td>0.00157</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>vrl</td>\n",
       "      <td>0</td>\n",
       "      <td>100887</td>\n",
       "      <td>22831</td>\n",
       "      <td>Soil-borne cereal mosaic virus</td>\n",
       "      <td>0.02816</td>\n",
       "      <td>0.01371</td>\n",
       "      <td>0.00767</td>\n",
       "      <td>0.03679</td>\n",
       "      <td>0.0138</td>\n",
       "      <td>0.00548</td>\n",
       "      <td>0.00473</td>\n",
       "      <td>0.02076</td>\n",
       "      <td>0.02716</td>\n",
       "      <td>0.00867</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.02773</td>\n",
       "      <td>0.02803</td>\n",
       "      <td>0.00508</td>\n",
       "      <td>0.0092</td>\n",
       "      <td>0.02965</td>\n",
       "      <td>0.02878</td>\n",
       "      <td>0.00574</td>\n",
       "      <td>0.01572</td>\n",
       "      <td>0.01577</td>\n",
       "      <td>0.01007</td>\n",
       "      <td>0.00508</td>\n",
       "      <td>0.00604</td>\n",
       "      <td>0.00679</td>\n",
       "      <td>0.01205</td>\n",
       "      <td>0.03127</td>\n",
       "      <td>0.00775</td>\n",
       "      <td>0.00959</td>\n",
       "      <td>0.00797</td>\n",
       "      <td>0.02006</td>\n",
       "      <td>0.00359</td>\n",
       "      <td>0.00933</td>\n",
       "      <td>0.01191</td>\n",
       "      <td>0.01616</td>\n",
       "      <td>0.00788</td>\n",
       "      <td>0.02593</td>\n",
       "      <td>0.00854</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.02098</td>\n",
       "      <td>0.02089</td>\n",
       "      <td>0.01367</td>\n",
       "      <td>0.01502</td>\n",
       "      <td>0.01809</td>\n",
       "      <td>0.02738</td>\n",
       "      <td>0.01796</td>\n",
       "      <td>0.01082</td>\n",
       "      <td>0.00705</td>\n",
       "      <td>0.01174</td>\n",
       "      <td>0.00858</td>\n",
       "      <td>0.03408</td>\n",
       "      <td>0.03964</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>0.00429</td>\n",
       "      <td>0.00578</td>\n",
       "      <td>0.00604</td>\n",
       "      <td>0.01494</td>\n",
       "      <td>0.01734</td>\n",
       "      <td>0.04148</td>\n",
       "      <td>0.02483</td>\n",
       "      <td>0.03359</td>\n",
       "      <td>0.03679</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00044</td>\n",
       "      <td>0.00131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>vrl</td>\n",
       "      <td>0</td>\n",
       "      <td>101029</td>\n",
       "      <td>5274</td>\n",
       "      <td>Human adenovirus type 7d</td>\n",
       "      <td>0.02579</td>\n",
       "      <td>0.02218</td>\n",
       "      <td>0.01479</td>\n",
       "      <td>0.01024</td>\n",
       "      <td>0.02294</td>\n",
       "      <td>0.00758</td>\n",
       "      <td>0.01782</td>\n",
       "      <td>0.01403</td>\n",
       "      <td>0.02636</td>\n",
       "      <td>0.01327</td>\n",
       "      <td>0.01896</td>\n",
       "      <td>0.02579</td>\n",
       "      <td>0.01877</td>\n",
       "      <td>0.01346</td>\n",
       "      <td>0.00721</td>\n",
       "      <td>0.01782</td>\n",
       "      <td>0.02067</td>\n",
       "      <td>0.02313</td>\n",
       "      <td>0.01214</td>\n",
       "      <td>0.00265</td>\n",
       "      <td>0.01232</td>\n",
       "      <td>0.01953</td>\n",
       "      <td>0.02105</td>\n",
       "      <td>0.00322</td>\n",
       "      <td>0.01232</td>\n",
       "      <td>0.01631</td>\n",
       "      <td>0.01327</td>\n",
       "      <td>0.0256</td>\n",
       "      <td>0.01119</td>\n",
       "      <td>0.0201</td>\n",
       "      <td>0.01763</td>\n",
       "      <td>0.00986</td>\n",
       "      <td>0.0036</td>\n",
       "      <td>0.00929</td>\n",
       "      <td>0.01138</td>\n",
       "      <td>0.03527</td>\n",
       "      <td>0.03015</td>\n",
       "      <td>0.02844</td>\n",
       "      <td>0.00284</td>\n",
       "      <td>0.01555</td>\n",
       "      <td>0.02825</td>\n",
       "      <td>0.01953</td>\n",
       "      <td>0.01251</td>\n",
       "      <td>0.03906</td>\n",
       "      <td>0.03546</td>\n",
       "      <td>0.01138</td>\n",
       "      <td>0.00948</td>\n",
       "      <td>0.00683</td>\n",
       "      <td>0.01043</td>\n",
       "      <td>0.03223</td>\n",
       "      <td>0.00986</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>0.00853</td>\n",
       "      <td>0.00322</td>\n",
       "      <td>0.00303</td>\n",
       "      <td>0.01593</td>\n",
       "      <td>0.00171</td>\n",
       "      <td>0.02427</td>\n",
       "      <td>0.02503</td>\n",
       "      <td>0.02825</td>\n",
       "      <td>0.0127</td>\n",
       "      <td>0.00133</td>\n",
       "      <td>0.00038</td>\n",
       "      <td>0.00209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>vrl</td>\n",
       "      <td>0</td>\n",
       "      <td>101688</td>\n",
       "      <td>3042</td>\n",
       "      <td>Apple latent spherical virus</td>\n",
       "      <td>0.04635</td>\n",
       "      <td>0.01545</td>\n",
       "      <td>0.02005</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.02761</td>\n",
       "      <td>0.01611</td>\n",
       "      <td>0.01052</td>\n",
       "      <td>0.00493</td>\n",
       "      <td>0.02597</td>\n",
       "      <td>0.00888</td>\n",
       "      <td>0.01512</td>\n",
       "      <td>0.02268</td>\n",
       "      <td>0.02893</td>\n",
       "      <td>0.00789</td>\n",
       "      <td>0.01151</td>\n",
       "      <td>0.01611</td>\n",
       "      <td>0.04011</td>\n",
       "      <td>0.01151</td>\n",
       "      <td>0.01118</td>\n",
       "      <td>0.00427</td>\n",
       "      <td>0.02794</td>\n",
       "      <td>0.01085</td>\n",
       "      <td>0.01216</td>\n",
       "      <td>0.00131</td>\n",
       "      <td>0.01216</td>\n",
       "      <td>0.02137</td>\n",
       "      <td>0.01052</td>\n",
       "      <td>0.01512</td>\n",
       "      <td>0.00986</td>\n",
       "      <td>0.03846</td>\n",
       "      <td>0.01085</td>\n",
       "      <td>0.01249</td>\n",
       "      <td>0.00066</td>\n",
       "      <td>0.01085</td>\n",
       "      <td>0.00427</td>\n",
       "      <td>0.02728</td>\n",
       "      <td>0.01381</td>\n",
       "      <td>0.01315</td>\n",
       "      <td>0.00427</td>\n",
       "      <td>0.02465</td>\n",
       "      <td>0.01151</td>\n",
       "      <td>0.02465</td>\n",
       "      <td>0.01085</td>\n",
       "      <td>0.03452</td>\n",
       "      <td>0.01348</td>\n",
       "      <td>0.01545</td>\n",
       "      <td>0.0069</td>\n",
       "      <td>0.01512</td>\n",
       "      <td>0.00756</td>\n",
       "      <td>0.02926</td>\n",
       "      <td>0.02696</td>\n",
       "      <td>0.00888</td>\n",
       "      <td>0.00625</td>\n",
       "      <td>0.0069</td>\n",
       "      <td>0.00329</td>\n",
       "      <td>0.01315</td>\n",
       "      <td>0.00822</td>\n",
       "      <td>0.04011</td>\n",
       "      <td>0.01183</td>\n",
       "      <td>0.02663</td>\n",
       "      <td>0.02663</td>\n",
       "      <td>0.00033</td>\n",
       "      <td>0.00033</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>vrl</td>\n",
       "      <td>0</td>\n",
       "      <td>101764</td>\n",
       "      <td>2801</td>\n",
       "      <td>Aconitum latent virus</td>\n",
       "      <td>0.02285</td>\n",
       "      <td>0.02678</td>\n",
       "      <td>0.01214</td>\n",
       "      <td>0.02321</td>\n",
       "      <td>0.01714</td>\n",
       "      <td>0.02213</td>\n",
       "      <td>0.00893</td>\n",
       "      <td>0.01928</td>\n",
       "      <td>0.01785</td>\n",
       "      <td>0.02356</td>\n",
       "      <td>0.01107</td>\n",
       "      <td>0.01999</td>\n",
       "      <td>0.01714</td>\n",
       "      <td>0.01142</td>\n",
       "      <td>0.00893</td>\n",
       "      <td>0.02499</td>\n",
       "      <td>0.03106</td>\n",
       "      <td>0.01357</td>\n",
       "      <td>0.01571</td>\n",
       "      <td>0.01321</td>\n",
       "      <td>0.01392</td>\n",
       "      <td>0.01107</td>\n",
       "      <td>0.01178</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.01142</td>\n",
       "      <td>0.01964</td>\n",
       "      <td>0.01714</td>\n",
       "      <td>0.01214</td>\n",
       "      <td>0.01464</td>\n",
       "      <td>0.01499</td>\n",
       "      <td>0.00536</td>\n",
       "      <td>0.00821</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.01678</td>\n",
       "      <td>0.01464</td>\n",
       "      <td>0.01535</td>\n",
       "      <td>0.01142</td>\n",
       "      <td>0.01607</td>\n",
       "      <td>0.01071</td>\n",
       "      <td>0.01142</td>\n",
       "      <td>0.02321</td>\n",
       "      <td>0.01856</td>\n",
       "      <td>0.01107</td>\n",
       "      <td>0.0282</td>\n",
       "      <td>0.01535</td>\n",
       "      <td>0.01142</td>\n",
       "      <td>0.01821</td>\n",
       "      <td>0.01071</td>\n",
       "      <td>0.01464</td>\n",
       "      <td>0.02785</td>\n",
       "      <td>0.02713</td>\n",
       "      <td>0.01499</td>\n",
       "      <td>0.01607</td>\n",
       "      <td>0.00714</td>\n",
       "      <td>0.00678</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>0.01107</td>\n",
       "      <td>0.03534</td>\n",
       "      <td>0.01571</td>\n",
       "      <td>0.03642</td>\n",
       "      <td>0.02785</td>\n",
       "      <td>0.00107</td>\n",
       "      <td>0.00036</td>\n",
       "      <td>0.00071</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Kingdom DNAtype SpeciesID Ncodons                              SpeciesName  \\\n",
       "0     vrl       0    100217    1995  Epizootic haematopoietic necrosis virus   \n",
       "1     vrl       0    100220    1474                         Bohle iridovirus   \n",
       "2     vrl       0    100755    4862             Sweet potato leaf curl virus   \n",
       "3     vrl       0    100880    1915             Northern cereal mosaic virus   \n",
       "4     vrl       0    100887   22831           Soil-borne cereal mosaic virus   \n",
       "5     vrl       0    101029    5274                 Human adenovirus type 7d   \n",
       "6     vrl       0    101688    3042             Apple latent spherical virus   \n",
       "7     vrl       0    101764    2801                    Aconitum latent virus   \n",
       "\n",
       "       UUU      UUC      UUA      UUG      CUU      CUC      CUA      CUG  \\\n",
       "0  0.01654  0.01203    5e-04  0.00351  0.01203  0.03208    0.001   0.0401   \n",
       "1  0.02714  0.01357  0.00068  0.00678  0.00407  0.02849  0.00204   0.0441   \n",
       "2  0.01974   0.0218  0.01357  0.01543  0.00782  0.01111  0.01028  0.01193   \n",
       "3  0.01775  0.02245  0.01619  0.00992  0.01567  0.01358   0.0094  0.01723   \n",
       "4  0.02816  0.01371  0.00767  0.03679   0.0138  0.00548  0.00473  0.02076   \n",
       "5  0.02579  0.02218  0.01479  0.01024  0.02294  0.00758  0.01782  0.01403   \n",
       "6  0.04635  0.01545  0.02005    0.024  0.02761  0.01611  0.01052  0.00493   \n",
       "7  0.02285  0.02678  0.01214  0.02321  0.01714  0.02213  0.00893  0.01928   \n",
       "\n",
       "       AUU      AUC      AUA      AUG      GUU      GUC      GUA      GUG  \\\n",
       "0  0.00551  0.02005  0.00752  0.02506  0.01103   0.0411  0.00902  0.03308   \n",
       "1  0.01153   0.0251  0.00882  0.03324  0.00814  0.04071  0.00814  0.03256   \n",
       "2  0.02283  0.01604  0.01316   0.0218  0.01625  0.01872  0.01213   0.0107   \n",
       "3  0.02402  0.02245  0.02507  0.02924  0.02089  0.02141  0.01723  0.01932   \n",
       "4  0.02716  0.00867   0.0131  0.02773  0.02803  0.00508   0.0092  0.02965   \n",
       "5  0.02636  0.01327  0.01896  0.02579  0.01877  0.01346  0.00721  0.01782   \n",
       "6  0.02597  0.00888  0.01512  0.02268  0.02893  0.00789  0.01151  0.01611   \n",
       "7  0.01785  0.02356  0.01107  0.01999  0.01714  0.01142  0.00893  0.02499   \n",
       "\n",
       "       GCU      GCC      GCA      GCG      CCU      CCC      CCA      CCG  \\\n",
       "0  0.01003  0.05013  0.01554  0.01103  0.02356  0.03208  0.01203  0.00501   \n",
       "1  0.01085  0.04885  0.01221  0.01357  0.00678  0.02714  0.01221  0.00407   \n",
       "2  0.02406  0.01234   0.0144  0.00514  0.01604   0.0146  0.02098   0.0107   \n",
       "3  0.02141  0.00679  0.02245  0.00522  0.01358  0.00418   0.0141  0.00574   \n",
       "4  0.02878  0.00574  0.01572  0.01577  0.01007  0.00508  0.00604  0.00679   \n",
       "5  0.02067  0.02313  0.01214  0.00265  0.01232  0.01953  0.02105  0.00322   \n",
       "6  0.04011  0.01151  0.01118  0.00427  0.02794  0.01085  0.01216  0.00131   \n",
       "7  0.03106  0.01357  0.01571  0.01321  0.01392  0.01107  0.01178    0.005   \n",
       "\n",
       "       UGG      GGU      GGC      GGA      GGG      UCU      UCC      UCA  \\\n",
       "0  0.01003  0.01203  0.03158  0.01905  0.02456  0.01353  0.02155  0.00251   \n",
       "1  0.01425  0.01221  0.01967  0.02239  0.01289  0.02103  0.01493  0.00407   \n",
       "2  0.01728  0.01851  0.00864  0.01172  0.01892  0.01933  0.01419  0.01296   \n",
       "3  0.01201  0.00992  0.00366  0.02402  0.02663  0.02872  0.00992   0.0235   \n",
       "4  0.01205  0.03127  0.00775  0.00959  0.00797  0.02006  0.00359  0.00933   \n",
       "5  0.01232  0.01631  0.01327   0.0256  0.01119   0.0201  0.01763  0.00986   \n",
       "6  0.01216  0.02137  0.01052  0.01512  0.00986  0.03846  0.01085  0.01249   \n",
       "7  0.01142  0.01964  0.01714  0.01214  0.01464  0.01499  0.00536  0.00821   \n",
       "\n",
       "       UCG      AGU      AGC      ACU      ACC      ACA      ACG      UAU  \\\n",
       "0  0.00652   0.0015  0.01554  0.00501  0.02105  0.00902  0.01053  0.00501   \n",
       "1  0.00475  0.00068  0.02035   0.0095  0.02782  0.01425  0.00611  0.00475   \n",
       "2  0.00967  0.01337  0.01337  0.01851  0.01131  0.01419   0.0109  0.02612   \n",
       "3  0.00522  0.01619  0.00836  0.02037  0.01358  0.02089  0.00731  0.02141   \n",
       "4  0.01191  0.01616  0.00788  0.02593  0.00854    0.012  0.02098  0.02089   \n",
       "5   0.0036  0.00929  0.01138  0.03527  0.03015  0.02844  0.00284  0.01555   \n",
       "6  0.00066  0.01085  0.00427  0.02728  0.01381  0.01315  0.00427  0.02465   \n",
       "7    0.005  0.01678  0.01464  0.01535  0.01142  0.01607  0.01071  0.01142   \n",
       "\n",
       "       UAC      CAA      CAG      AAU      AAC      UGU      UGC      CAU  \\\n",
       "0  0.02256  0.00301  0.03108  0.00401  0.02607  0.00251  0.01153  0.00501   \n",
       "1  0.02917  0.00407  0.02374  0.00882  0.02917  0.00271  0.01628  0.00204   \n",
       "2  0.01275  0.01522  0.02365  0.02962  0.01789  0.01625  0.01234  0.01604   \n",
       "3  0.00888  0.01567  0.01253  0.02298  0.01358  0.00992  0.00888  0.00783   \n",
       "4  0.01367  0.01502  0.01809  0.02738  0.01796  0.01082  0.00705  0.01174   \n",
       "5  0.02825  0.01953  0.01251  0.03906  0.03546  0.01138  0.00948  0.00683   \n",
       "6  0.01151  0.02465  0.01085  0.03452  0.01348  0.01545   0.0069  0.01512   \n",
       "7  0.02321  0.01856  0.01107   0.0282  0.01535  0.01142  0.01821  0.01071   \n",
       "\n",
       "       CAC      AAA      AAG      CGU      CGC      CGA      CGG      AGA  \\\n",
       "0  0.02356  0.01053   0.0386  0.00401  0.00702  0.00401  0.00451  0.01303   \n",
       "1  0.01967  0.00543  0.03392  0.00136  0.00678  0.00136  0.00136  0.01696   \n",
       "2  0.01687  0.02077  0.03949  0.00864  0.00596  0.00926  0.00596  0.01974   \n",
       "3  0.00679  0.03133  0.04282  0.00627  0.00261  0.00261  0.00366   0.0141   \n",
       "4  0.00858  0.03408  0.03964   0.0095  0.00429  0.00578  0.00604  0.01494   \n",
       "5  0.01043  0.03223  0.00986  0.00398  0.00853  0.00322  0.00303  0.01593   \n",
       "6  0.00756  0.02926  0.02696  0.00888  0.00625   0.0069  0.00329  0.01315   \n",
       "7  0.01464  0.02785  0.02713  0.01499  0.01607  0.00714  0.00678   0.0125   \n",
       "\n",
       "       AGG      GAU      GAC      GAA      GAG      UAA      UAG      UGA  \n",
       "0  0.03559  0.01003  0.04612  0.01203  0.04361  0.00251    5e-04        0  \n",
       "1  0.03596  0.01221  0.04545   0.0156   0.0441  0.00271  0.00068        0  \n",
       "2  0.02489  0.03126  0.02036  0.02242  0.02468  0.00391        0  0.00144  \n",
       "3  0.01671   0.0376  0.01932  0.03029  0.03446  0.00261  0.00157        0  \n",
       "4  0.01734  0.04148  0.02483  0.03359  0.03679        0  0.00044  0.00131  \n",
       "5  0.00171  0.02427  0.02503  0.02825   0.0127  0.00133  0.00038  0.00209  \n",
       "6  0.00822  0.04011  0.01183  0.02663  0.02663  0.00033  0.00033        0  \n",
       "7  0.01107  0.03534  0.01571  0.03642  0.02785  0.00107  0.00036  0.00071  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Heuristic by name found -> codon_col: Ncodons aa_col: CAA freq_col: None\n",
      "\n",
      "Top codon-like column scores (col:score):\n",
      "  'Kingdom': 0.000\n",
      "  'DNAtype': 0.000\n",
      "  'SpeciesID': 0.000\n",
      "  'Ncodons': 0.000\n",
      "  'SpeciesName': 0.000\n",
      "  'UUU': 0.000\n",
      "  'UUC': 0.000\n",
      "  'UUA': 0.000\n",
      "\n",
      "Using detected codon column -> 'Ncodons'\n",
      "\n",
      "ERROR: Parsing produced no rows. Please paste df_raw.head(10) output here so I can inspect file format.\n"
     ]
    }
   ],
   "source": [
    "# Robust single-cell parser + diagnostics for codon_usage.csv\n",
    "import pandas as pd\n",
    "import re\n",
    "from Bio.Seq import Seq\n",
    "from collections import Counter, defaultdict\n",
    "import os\n",
    "\n",
    "CSV_PATH = \"codon_usage.csv\"   # change if needed\n",
    "if not os.path.exists(CSV_PATH):\n",
    "    raise FileNotFoundError(f\"File not found: {CSV_PATH} - put the CSV in the notebook folder or change path.\")\n",
    "\n",
    "# 1) Load file (all as str) and show basic info\n",
    "df_raw = pd.read_csv(CSV_PATH, dtype=str)\n",
    "pd.set_option('display.max_columns', 200)\n",
    "print(\"Loaded:\", CSV_PATH)\n",
    "print(\"Shape:\", df_raw.shape)\n",
    "print(\"Columns:\")\n",
    "for i,c in enumerate(df_raw.columns):\n",
    "    print(f\"  {i+1}. {c!r}\")\n",
    "print(\"\\nFirst 8 rows preview:\")\n",
    "display(df_raw.head(8))\n",
    "\n",
    "# helper tests\n",
    "def is_codon_token(s):\n",
    "    if not isinstance(s, str): return False\n",
    "    s = s.strip().upper().replace(\"U\",\"T\")\n",
    "    return bool(re.fullmatch(r\"[ACGT]{3}\", s))\n",
    "\n",
    "def column_codon_score(series, ncheck=500):\n",
    "    total=0; codon_like=0\n",
    "    for v in series.dropna().astype(str).head(ncheck):\n",
    "        total+=1\n",
    "        if is_codon_token(v): codon_like+=1\n",
    "    return (codon_like/total) if total>0 else 0.0\n",
    "\n",
    "def column_aa_score(series, ncheck=500):\n",
    "    total=0; aa_like=0\n",
    "    allowed = set(list(\"ACDEFGHIKLMNPQRSTVWY\"))\n",
    "    for v in series.dropna().astype(str).head(ncheck):\n",
    "        vv=v.strip().upper()\n",
    "        total+=1\n",
    "        if len(vv)==1 and vv in allowed:\n",
    "            aa_like+=1\n",
    "    return (aa_like/total) if total>0 else 0.0\n",
    "\n",
    "# 2) Heuristic: find columns by name\n",
    "codon_col=None; aa_col=None; freq_col=None\n",
    "for c in df_raw.columns:\n",
    "    lc=c.lower()\n",
    "    if any(k in lc for k in (\"codon\",\"triplet\",\"trip\",\"tri\",\"triplet\",\"tripletcodon\")) and codon_col is None:\n",
    "        codon_col=c\n",
    "    if any(k in lc for k in (\"amino\",\"aa\",\"residue\")) and aa_col is None:\n",
    "        aa_col=c\n",
    "    if any(k in lc for k in (\"freq\",\"frequency\",\"usage\",\"percent\",\"count\")) and freq_col is None:\n",
    "        freq_col=c\n",
    "\n",
    "print(\"\\nHeuristic by name found -> codon_col:\", codon_col, \"aa_col:\", aa_col, \"freq_col:\", freq_col)\n",
    "\n",
    "# 3) Score columns by codon-like content (print top scored)\n",
    "scores = {c: column_codon_score(df_raw[c]) for c in df_raw.columns}\n",
    "scores_sorted = sorted(scores.items(), key=lambda x:-x[1])\n",
    "print(\"\\nTop codon-like column scores (col:score):\")\n",
    "for c,s in scores_sorted[:8]:\n",
    "    print(f\"  {c!r}: {s:.3f}\")\n",
    "\n",
    "# 4) If we didn't find codon_col by name, pick top-scoring column if score >= threshold\n",
    "if not codon_col:\n",
    "    best_col, best_score = scores_sorted[0]\n",
    "    if best_score >= 0.2:\n",
    "        codon_col = best_col\n",
    "        print(f\"\\nAuto-selected codon column: {codon_col} with score {best_score:.3f}\")\n",
    "    else:\n",
    "        print(\"\\nNo column has strong codon-like content (score <0.2). We'll scan ALL cells for 3-letter tokens and build rows.\")\n",
    "\n",
    "# 5) If aa_col is missing, score for amino-acid-like columns\n",
    "if not aa_col:\n",
    "    aa_scores = {c: column_aa_score(df_raw[c]) for c in df_raw.columns}\n",
    "    aa_sorted = sorted(aa_scores.items(), key=lambda x:-x[1])\n",
    "    print(\"\\nTop amino-acid-like column scores (col:score):\")\n",
    "    for c,s in aa_sorted[:8]:\n",
    "        print(f\"  {c!r}: {s:.3f}\")\n",
    "    # pick if >= 0.15\n",
    "    if aa_sorted and aa_sorted[0][1] >= 0.15:\n",
    "        aa_col = aa_sorted[0][0]\n",
    "        print(f\"\\nAuto-selected amino-acid column: {aa_col} with score {aa_sorted[0][1]:.3f}\")\n",
    "\n",
    "# 6) Build parsed rows using best available info OR scan all cells for codon tokens\n",
    "rows=[]\n",
    "if codon_col:\n",
    "    print(f\"\\nUsing detected codon column -> {codon_col!r}\")\n",
    "    for _, r in df_raw.iterrows():\n",
    "        val = r.get(codon_col, \"\")\n",
    "        if not isinstance(val, str): continue\n",
    "        codon = val.strip().upper().replace(\"U\",\"T\")\n",
    "        if not is_codon_token(codon): \n",
    "            continue\n",
    "        # find aa\n",
    "        aa_val = None\n",
    "        if aa_col:\n",
    "            aa_val = str(r.get(aa_col,\"\")).strip().upper()\n",
    "            if not (isinstance(aa_val,str) and len(aa_val)==1 and aa_val in list(\"ACDEFGHIKLMNPQRSTVWY\")):\n",
    "                aa_val = None\n",
    "        if not aa_val:\n",
    "            # try translate codon\n",
    "            try:\n",
    "                aa_val = str(Seq(codon).translate())\n",
    "            except:\n",
    "                aa_val = \"X\"\n",
    "        rows.append({\"aa\": aa_val, \"orig_codon\": codon, \"orig_freq\": r.get(freq_col, None)})\n",
    "else:\n",
    "    # scan every cell for codon tokens; aggregate counts and record parent row index\n",
    "    print(\"\\nScanning all cells for codon tokens (may produce many rows)...\")\n",
    "    for idx in range(min(len(df_raw), 5000)):  # limit scans to first 5000 rows for speed\n",
    "        row = df_raw.iloc[idx]\n",
    "        for c in df_raw.columns:\n",
    "            v = row[c]\n",
    "            if not isinstance(v, str): \n",
    "                continue\n",
    "            v2 = v.strip().upper().replace(\"U\",\"T\")\n",
    "            if is_codon_token(v2):\n",
    "                # translate\n",
    "                try:\n",
    "                    aa_val = str(Seq(v2).translate())\n",
    "                except:\n",
    "                    aa_val = \"X\"\n",
    "                rows.append({\"aa\": aa_val, \"orig_codon\": v2, \"orig_freq\": None, \"source_row\": idx, \"source_col\": c})\n",
    "\n",
    "# 7) Create DataFrame and save\n",
    "if not rows:\n",
    "    print(\"\\nERROR: Parsing produced no rows. Please paste df_raw.head(10) output here so I can inspect file format.\")\n",
    "else:\n",
    "    parsed = pd.DataFrame(rows)\n",
    "    # normalize columns\n",
    "    parsed = parsed.rename(columns={c:c for c in parsed.columns})\n",
    "    # ensure columns exist\n",
    "    if 'aa' not in parsed.columns or 'orig_codon' not in parsed.columns:\n",
    "        print(\"Parsed DataFrame missing required columns. Columns:\", parsed.columns)\n",
    "    else:\n",
    "        print(\"\\nParsed rows count:\", parsed.shape[0])\n",
    "        print(\"Unique codons:\", parsed['orig_codon'].nunique())\n",
    "        print(\"Unique amino acids:\", parsed['aa'].nunique())\n",
    "        display(parsed.head(20))\n",
    "        out = \"codon_parsed_from_kaggle.csv\"\n",
    "        parsed.to_csv(out, index=False)\n",
    "        print(f\"\\nSaved parsed CSV to: {out} (you can now continue training pipeline using this file)\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a88118fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Kingdom', 'DNAtype', 'SpeciesID', 'Ncodons', 'SpeciesName', 'UUU', 'UUC', 'UUA', 'UUG', 'CUU', 'CUC', 'CUA', 'CUG', 'AUU', 'AUC', 'AUA', 'AUG', 'GUU', 'GUC', 'GUA', 'GUG', 'GCU', 'GCC', 'GCA', 'GCG', 'CCU', 'CCC', 'CCA', 'CCG', 'UGG', 'GGU', 'GGC', 'GGA', 'GGG', 'UCU', 'UCC', 'UCA', 'UCG', 'AGU', 'AGC', 'ACU', 'ACC', 'ACA', 'ACG', 'UAU', 'UAC', 'CAA', 'CAG', 'AAU', 'AAC', 'UGU', 'UGC', 'CAU', 'CAC', 'AAA', 'AAG', 'CGU', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG', 'GAU', 'GAC', 'GAA', 'GAG', 'UAA', 'UAG', 'UGA']\n",
      "Kingdom DNAtype SpeciesID Ncodons                             SpeciesName     UUU     UUC     UUA     UUG     CUU     CUC     CUA     CUG     AUU     AUC     AUA     AUG     GUU     GUC     GUA     GUG     GCU     GCC     GCA     GCG     CCU     CCC     CCA     CCG     UGG     GGU     GGC     GGA     GGG     UCU     UCC     UCA     UCG     AGU     AGC     ACU     ACC     ACA     ACG     UAU     UAC     CAA     CAG     AAU     AAC     UGU     UGC     CAU     CAC     AAA     AAG     CGU     CGC     CGA     CGG     AGA     AGG     GAU     GAC     GAA     GAG     UAA     UAG     UGA\n",
      "    vrl       0    100217    1995 Epizootic haematopoietic necrosis virus 0.01654 0.01203   5e-04 0.00351 0.01203 0.03208   0.001  0.0401 0.00551 0.02005 0.00752 0.02506 0.01103  0.0411 0.00902 0.03308 0.01003 0.05013 0.01554 0.01103 0.02356 0.03208 0.01203 0.00501 0.01003 0.01203 0.03158 0.01905 0.02456 0.01353 0.02155 0.00251 0.00652  0.0015 0.01554 0.00501 0.02105 0.00902 0.01053 0.00501 0.02256 0.00301 0.03108 0.00401 0.02607 0.00251 0.01153 0.00501 0.02356 0.01053  0.0386 0.00401 0.00702 0.00401 0.00451 0.01303 0.03559 0.01003 0.04612 0.01203 0.04361 0.00251   5e-04       0\n",
      "    vrl       0    100220    1474                        Bohle iridovirus 0.02714 0.01357 0.00068 0.00678 0.00407 0.02849 0.00204  0.0441 0.01153  0.0251 0.00882 0.03324 0.00814 0.04071 0.00814 0.03256 0.01085 0.04885 0.01221 0.01357 0.00678 0.02714 0.01221 0.00407 0.01425 0.01221 0.01967 0.02239 0.01289 0.02103 0.01493 0.00407 0.00475 0.00068 0.02035  0.0095 0.02782 0.01425 0.00611 0.00475 0.02917 0.00407 0.02374 0.00882 0.02917 0.00271 0.01628 0.00204 0.01967 0.00543 0.03392 0.00136 0.00678 0.00136 0.00136 0.01696 0.03596 0.01221 0.04545  0.0156  0.0441 0.00271 0.00068       0\n",
      "    vrl       0    100755    4862            Sweet potato leaf curl virus 0.01974  0.0218 0.01357 0.01543 0.00782 0.01111 0.01028 0.01193 0.02283 0.01604 0.01316  0.0218 0.01625 0.01872 0.01213  0.0107 0.02406 0.01234  0.0144 0.00514 0.01604  0.0146 0.02098  0.0107 0.01728 0.01851 0.00864 0.01172 0.01892 0.01933 0.01419 0.01296 0.00967 0.01337 0.01337 0.01851 0.01131 0.01419  0.0109 0.02612 0.01275 0.01522 0.02365 0.02962 0.01789 0.01625 0.01234 0.01604 0.01687 0.02077 0.03949 0.00864 0.00596 0.00926 0.00596 0.01974 0.02489 0.03126 0.02036 0.02242 0.02468 0.00391       0 0.00144\n",
      "    vrl       0    100880    1915            Northern cereal mosaic virus 0.01775 0.02245 0.01619 0.00992 0.01567 0.01358  0.0094 0.01723 0.02402 0.02245 0.02507 0.02924 0.02089 0.02141 0.01723 0.01932 0.02141 0.00679 0.02245 0.00522 0.01358 0.00418  0.0141 0.00574 0.01201 0.00992 0.00366 0.02402 0.02663 0.02872 0.00992  0.0235 0.00522 0.01619 0.00836 0.02037 0.01358 0.02089 0.00731 0.02141 0.00888 0.01567 0.01253 0.02298 0.01358 0.00992 0.00888 0.00783 0.00679 0.03133 0.04282 0.00627 0.00261 0.00261 0.00366  0.0141 0.01671  0.0376 0.01932 0.03029 0.03446 0.00261 0.00157       0\n",
      "    vrl       0    100887   22831          Soil-borne cereal mosaic virus 0.02816 0.01371 0.00767 0.03679  0.0138 0.00548 0.00473 0.02076 0.02716 0.00867  0.0131 0.02773 0.02803 0.00508  0.0092 0.02965 0.02878 0.00574 0.01572 0.01577 0.01007 0.00508 0.00604 0.00679 0.01205 0.03127 0.00775 0.00959 0.00797 0.02006 0.00359 0.00933 0.01191 0.01616 0.00788 0.02593 0.00854   0.012 0.02098 0.02089 0.01367 0.01502 0.01809 0.02738 0.01796 0.01082 0.00705 0.01174 0.00858 0.03408 0.03964  0.0095 0.00429 0.00578 0.00604 0.01494 0.01734 0.04148 0.02483 0.03359 0.03679       0 0.00044 0.00131\n",
      "    vrl       0    101029    5274                Human adenovirus type 7d 0.02579 0.02218 0.01479 0.01024 0.02294 0.00758 0.01782 0.01403 0.02636 0.01327 0.01896 0.02579 0.01877 0.01346 0.00721 0.01782 0.02067 0.02313 0.01214 0.00265 0.01232 0.01953 0.02105 0.00322 0.01232 0.01631 0.01327  0.0256 0.01119  0.0201 0.01763 0.00986  0.0036 0.00929 0.01138 0.03527 0.03015 0.02844 0.00284 0.01555 0.02825 0.01953 0.01251 0.03906 0.03546 0.01138 0.00948 0.00683 0.01043 0.03223 0.00986 0.00398 0.00853 0.00322 0.00303 0.01593 0.00171 0.02427 0.02503 0.02825  0.0127 0.00133 0.00038 0.00209\n",
      "    vrl       0    101688    3042            Apple latent spherical virus 0.04635 0.01545 0.02005   0.024 0.02761 0.01611 0.01052 0.00493 0.02597 0.00888 0.01512 0.02268 0.02893 0.00789 0.01151 0.01611 0.04011 0.01151 0.01118 0.00427 0.02794 0.01085 0.01216 0.00131 0.01216 0.02137 0.01052 0.01512 0.00986 0.03846 0.01085 0.01249 0.00066 0.01085 0.00427 0.02728 0.01381 0.01315 0.00427 0.02465 0.01151 0.02465 0.01085 0.03452 0.01348 0.01545  0.0069 0.01512 0.00756 0.02926 0.02696 0.00888 0.00625  0.0069 0.00329 0.01315 0.00822 0.04011 0.01183 0.02663 0.02663 0.00033 0.00033       0\n",
      "    vrl       0    101764    2801                   Aconitum latent virus 0.02285 0.02678 0.01214 0.02321 0.01714 0.02213 0.00893 0.01928 0.01785 0.02356 0.01107 0.01999 0.01714 0.01142 0.00893 0.02499 0.03106 0.01357 0.01571 0.01321 0.01392 0.01107 0.01178   0.005 0.01142 0.01964 0.01714 0.01214 0.01464 0.01499 0.00536 0.00821   0.005 0.01678 0.01464 0.01535 0.01142 0.01607 0.01071 0.01142 0.02321 0.01856 0.01107  0.0282 0.01535 0.01142 0.01821 0.01071 0.01464 0.02785 0.02713 0.01499 0.01607 0.00714 0.00678  0.0125 0.01107 0.03534 0.01571 0.03642 0.02785 0.00107 0.00036 0.00071\n",
      "    vrl       0    101947    2897                   Pseudorabies virus Ea 0.01105 0.02106 0.00035 0.00104 0.00035 0.04142 0.00069 0.05868 0.00104 0.01933 0.00173 0.01691       0 0.04004 0.00069 0.04522 0.00276 0.06973 0.00069 0.05212 0.00173 0.04764 0.00173 0.04177 0.01139 0.00242 0.04108 0.00621 0.02969 0.00449 0.01622 0.00035 0.02313 0.00035 0.02071 0.00104 0.02589       0 0.03176 0.00173 0.02485 0.00138 0.01622 0.00138 0.01519 0.00035 0.01968 0.00173 0.01933 0.00035 0.00897 0.00207  0.0535 0.00311 0.02658 0.00207 0.00311 0.00414 0.04556 0.00449 0.04867 0.00138 0.00035 0.00138\n",
      "    vrl       0     10249   61247               Vaccinia virus Copenhagen 0.03411  0.0143 0.02771 0.01869 0.01148 0.00651  0.0186  0.0073 0.03713 0.01915 0.03896 0.02856 0.02025 0.00878 0.02338 0.00916 0.01264 0.00671 0.01133 0.00718 0.01097 0.00423 0.01239 0.00624 0.00731 0.01246 0.00371  0.0202 0.00271 0.02527 0.01053   0.016  0.0089 0.01489 0.00612 0.02158  0.0086 0.02111 0.00913 0.03509 0.01623 0.01533 0.00622  0.0466 0.02077 0.01616 0.00503 0.01517 0.00625 0.04835 0.02222 0.00634 0.00248 0.00578 0.00167  0.0223 0.00411 0.04866 0.01559 0.03695 0.01412  0.0025 0.00077 0.00103\n",
      "Kingdom        object\n",
      "DNAtype        object\n",
      "SpeciesID      object\n",
      "Ncodons        object\n",
      "SpeciesName    object\n",
      "                ...  \n",
      "GAA            object\n",
      "GAG            object\n",
      "UAA            object\n",
      "UAG            object\n",
      "UGA            object\n",
      "Length: 69, dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(df_raw.columns.tolist())\n",
    "print(df_raw.head(10).to_string(index=False))\n",
    "print(df_raw.dtypes)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
